### PR TITLE
add `DataLoader` related parameters to `fit()` and `predict()`

### PR DESCRIPTION
**Pull request by BohdanBilonoh**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md). 

### Summary

Add `torch.utils.data.DataLoader` related parameters to `fit()` and `predict()` of `TorchForecastingModel`

### CODE REVIEW

**Comment by madtoinou** [comment]
Hi @BohdanBilonoh,

It looks great, however to make it easier to maintain and more exhaustive, I think that it would be great to just add an argument called `dataloader_kwargs`, then check that the argument explicitly used by Darts are not redundant/overwritten and then pass this argument down to the `DataLoader` constructor.

It will allow users to specify more than just `prefetch_factor`, `persistent_workers` and `pin_memory`, while limiting copy-pasting from other library documentation (putting a link to the `torch.DataLoader` page does sound like a good idea for this argument however)

**Comment by joshua-xia** [comment]
Hi @BohdanBilonoh, Would you please add multiprocessing_context parameter for Dataloader, it is useful when we use multi-workers for dataloader, Thanks!

**Comment by joshua-xia** [comment]
> Hi @BohdanBilonoh, Would you please add multiprocessing_context parameter for Dataloader, it is useful when we use multi-workers for dataloader, Thanks!

@BohdanBilonoh refer to #2375

**Comment by joshua-xia** [comment]
> > Hi @BohdanBilonoh, Would you please add multiprocessing_context parameter for Dataloader, it is useful when we use multi-workers for dataloader, Thanks!
> 
> @BohdanBilonoh refer to #2375

@BohdanBilonoh  My bad, it is good idea from @madtoinou to add dataloader_kwargs to let user input dataloader parameters as wish freely, not need to support multiprocessing_context parameter forcibly

**Comment by BohdanBilonoh** [comment]
@madtoinou what do you think about hardcoded parameters like

```python
batch_size=self.batch_size,
shuffle=True,
drop_last=False,
collate_fn=self._batch_collate_fn,
```

should it be hard coded for new `dataloader_kwargs` like

```python
def _setup_for_train(
    self,
    train_dataset: TrainingDataset,
    val_dataset: Optional[TrainingDataset] = None,
    trainer: Optional[pl.Trainer] = None,
    verbose: Optional[bool] = None,
    epochs: int = 0,
    dataloader_kwargs: Optional[Dict[str, Any]] = None,
) -> Tuple[pl.Trainer, PLForecastingModule, DataLoader, Optional[DataLoader]]:

        ...

        if dataloader_kwargs is None:
            dataloader_kwargs = {}

        dataloader_kwargs["shuffle"] = True
        dataloader_kwargs["batch_size"] = self.batch_size
        dataloader_kwargs["drop_last"] = False
        dataloader_kwargs["collate_fn"] = self._batch_collate_fn

        # Setting drop_last to False makes the model see each sample at least once, and guarantee the presence of at
        # least one batch no matter the chosen batch size
        train_loader = DataLoader(
            train_dataset,
            **dataloader_kwargs,
        )

        dataloader_kwargs["shuffle"] = False

        # Prepare validation data
        val_loader = (
            None
            if val_dataset is None
            else DataLoader(
                val_dataset,
                **dataloader_kwargs,
            )
        )
```

or give a user full control on `dataloader_kwargs`?

**Comment by tRosenflanz** [comment]
> @madtoinou what do you think about hardcoded parameters like
> 
> ```python
> batch_size=self.batch_size,
> shuffle=True,
> drop_last=False,
> collate_fn=self._batch_collate_fn,
> ```
> 
> should it be hard coded for new `dataloader_kwargs` like
> 
> ```python
> def _setup_for_train(
>     self,
>     train_dataset: TrainingDataset,
>     val_dataset: Optional[TrainingDataset] = None,
>     trainer: Optional[pl.Trainer] = None,
>     verbose: Optional[bool] = None,
>     epochs: int = 0,
>     dataloader_kwargs: Optional[Dict[str, Any]] = None,
> ) -> Tuple[pl.Trainer, PLForecastingModule, DataLoader, Optional[DataLoader]]:
> 
>         ...
> 
>         if dataloader_kwargs is None:
>             dataloader_kwargs = {}
> 
>         dataloader_kwargs["shuffle"] = True
>         dataloader_kwargs["batch_size"] = self.batch_size
>         dataloader_kwargs["drop_last"] = False
>         dataloader_kwargs["collate_fn"] = self._batch_collate_fn
> 
>         # Setting drop_last to False makes the model see each sample at least once, and guarantee the presence of at
>         # least one batch no matter the chosen batch size
>         train_loader = DataLoader(
>             train_dataset,
>             **dataloader_kwargs,
>         )
> 
>         dataloader_kwargs["shuffle"] = False
> 
>         # Prepare validation data
>         val_loader = (
>             None
>             if val_dataset is None
>             else DataLoader(
>                 val_dataset,
>                 **dataloader_kwargs,
>             )
>         )
> 
>         ...
> ```
> 
> or give a user full control on `dataloader_kwargs`?

you could extend your suggestion to allow overrides but with populated defaults 
```
defaults = dict(shuffle = True, batch_size = self.batch_size, drop_last = False, collate_fn = self._batch_collate_fn)
#combine with defaults but override them
dataloader_kwargs_train = {**defaults,**dataloader_kwargs}
#override shuffle
dataloader_kwargs_val = (**dataloader_kwargs_train, **dict(shuffle=False)} 
```

File: darts/models/forecasting/torch_forecasting_model.py [hunk 1]
```diff
@@ -1487,14 +1484,17 @@ def predict_from_dataset(
             mc_dropout=mc_dropout,
         )
 
+        if data_loader_kwargs is None:
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py:1487]: Shouldn't we allow then here as well the liberty to overwrite these defaults (expect shuffle)?

**Review by dennisbader — CHANGES_REQUESTED** [review]
Thanks for the updates @BohdanBilonoh, I took the chance to fix some indentation issues in one of the docstrings and pushed the changes.

If we give the freedom to overwrite our default dataloader params when calling `fit()`, shouldn't we then also allow that during `predict()`?

Also, removing the `num_loader_workers` parameter is a breaking change. Can you document this in the CHANGELOG.md?

File: darts/models/forecasting/torch_forecasting_model.py [hunk 2]
```diff
@@ -1281,11 +1280,10 @@ def predict(
         num_samples
             Number of times a prediction is sampled from a probabilistic model. Should be left set to 1
             for deterministic models.
-        num_loader_workers
-            Optionally, an integer specifying the ``num_workers`` to use in PyTorch ``DataLoader`` instances,
-            for the inference/prediction dataset loaders (if any).
-            A larger number of workers can sometimes increase performance, but can also incur extra overheads
-            and increase memory usage, as more batches are loaded in parallel.
+        data_loader_kwargs
+            Optionally, a dictionary of keyword arguments to pass to the PyTorch DataLoader instances used to load the
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py:1284]: it's referring to the train and val datasets, but it should be for the prediction dataset, same for the methods below

**Review by dennisbader — APPROVED** [review]
Thanks for the changes @BohdanBilonoh. I made some last updates including unit tests and some docstrings.

Looks great now, and ready to be merged once tests have passed!

Thanks a lot for the great work 🚀

**Review by madtoinou — APPROVED** [review]
